### ISSUE TITLE
Two concurrent read and write processes can raise FileModeError on read close()

### ISSUE THREAD

**Issue by CharlesB2** [PyTables/PyTables#700]
I have two concurrent processes, one reading a table, and one background process writing a table with flushes every minute. Table is indexed, and autoindex set to True.

Once upon a time the read process raises `tables.exceptions.FileModeError: the file is not writable`.
Complete stack trace:
```
  File "/home/charles/venv/lib/python3.5/site-packages/tables/file.py", line 2771, in __exit__
    self.close()
  File "/home/charles/venv/lib/python3.5/site-packages/tables/file.py", line 2732, in close
    self.root._f_close()
  File "/home/charles/venv/lib/python3.5/site-packages/tables/group.py", line 937, in _f_close
    self._g_close_descendents()
  File "/home/charles/venv/lib/python3.5/site-packages/tables/group.py", line 901, in _g_close_descendents
    node_manager.close_subtree(self._v_pathname)
  File "/home/charles/venv/lib/python3.5/site-packages/tables/file.py", line 543, in close_subtree
    self._close_nodes(paths, cache.pop)
  File "/home/charles/venv/lib/python3.5/site-packages/tables/file.py", line 518, in _close_nodes
    node._f_close()
  File "/home/charles/venv/lib/python3.5/site-packages/tables/table.py", line 2965, in _f_close
    self.flush()
  File "/home/charles/venv/lib/python3.5/site-packages/tables/table.py", line 2902, in flush
    rowsadded = self.flush_rows_to_index(_lastrow=True)
  File "/home/charles/venv/lib/python3.5/site-packages/tables/table.py", line 2527, in flush_rows_to_index
    colname, start, nrows, _lastrow, update=True)
  File "/home/charles/venv/lib/python3.5/site-packages/tables/table.py", line 2557, in _add_rows_to_index
    update=update)
  File "/home/charles/venv/lib/python3.5/site-packages/tables/index.py", line 739, in append_last_row
    larr, arr, idx = self.initial_append(xarr, nrows, reduction)
  File "/home/charles/venv/lib/python3.5/site-packages/tables/index.py", line 631, in initial_append
    self._v_attrs.is_csi = False
  File "/home/charles/venv/lib/python3.5/site-packages/tables/attributeset.py", line 477, in __setattr__
    nodefile._check_writable()
  File "/home/charles/venv/lib/python3.5/site-packages/tables/file.py", line 2176, in _check_writable
    raise FileModeError("the file is not writable")
```

**Comment by CharlesB2** [comment]
The processes are on different hosts, the read process accesses the file through SSHFS, though quite slow.

The table is relatively large. I think that sometimes the read process opens the file while the write process has flushed table rows, but not flushed index yet (it runs on raspberry so it can be slow)

In `_g_post_init_hook()` the read process sees incomplete index: `self.nrows > self._indexedrows` (`tables/table.py, line 913), and will try to flush on close.

An easy fix would be to flush only if file is opened in read mode?

**Comment by CharlesB2** [comment]
An easy fix would be to flush only if file is opened in write or append mode?

**Comment by FrancescAlted** [comment]
PyTables does not support this concurrent read/write scenarios, historically due to HDF5 limitations.  We know that newest versions of HDF5 (e.g. 1.10.4) does have support for this, but PyTables has not implemented it yet.  This would be nice to have indeed and we would be happy to accept PR for this (asking for a small grant from NumFOCUS is a realistic option; any takers?).

**Comment by avalentino** [comment]
Closed in favor of #788.
